[jit] Add RRef to IValue and Jit

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32959 jit pickling rref
* **#32674 [jit] Add RRef to IValue and Jit**
* #32990 allow remote torchscript call to itself

Differential Revision: [D19502233](https://our.internmc.facebook.com/intern/diff/D19502233/)

zhaojuanmao committed:
[jit] Add RRef to IValue and Jit



Differential Revision: [D19502233](https://our.internmc.facebook.com/intern/diff/D19502233/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 9b63aa0a:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=9b63aa0a225bacd671a050bf91c26716d7304290).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4353050) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=91639547) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=91639547))
<details>
<summary>
<code>Jan 28 01:09:28 /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/init.cpp:376:34: error: no matching function for call to 'torch::distributed::rpc::PyRRef::PyRRef(std::shared_ptr<torch::distributed::rpc::UserRRef>&)'</code>
</summary>

```
Jan 28 01:09:28    AT_CHECK( 
Jan 28 01:09:28    ^ 
Jan 28 01:09:28 In file included from /var/lib/jenkins/workspace/c10/util/intrusive_ptr.h:4:0, 
Jan 28 01:09:28                  from /var/lib/jenkins/workspace/c10/core/thread_pool.h:11, 
Jan 28 01:09:28                  from /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/process_group_agent.h:3, 
Jan 28 01:09:28                  from /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/init.cpp:3: 
Jan 28 01:09:28 /var/lib/jenkins/workspace/c10/util/Exception.h:350:13: note: declared here 
Jan 28 01:09:28  inline void deprecated_AT_CHECK() {} 
Jan 28 01:09:28              ^ 
Jan 28 01:09:28 /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/init.cpp: In lambda function: 
Jan 28 01:09:28 /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/init.cpp:376:34: error: no matching function for call to 'torch::distributed::rpc::PyRRef::PyRRef(std::shared_ptr<torch::distributed::rpc::UserRRef>&)' 
Jan 28 01:09:28          return PyRRef(userRRefPtr); 
Jan 28 01:09:28                                   ^ 
Jan 28 01:09:28 In file included from /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/init.cpp:4:0: 
Jan 28 01:09:28 /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/py_rref.h:16:12: note: candidate: torch::distributed::rpc::PyRRef::PyRRef(c10::intrusive_ptr<torch::distributed::rpc::RRef>) 
Jan 28 01:09:28    explicit PyRRef(c10::intrusive_ptr<RRef> rref); 
Jan 28 01:09:28             ^ 
Jan 28 01:09:28 /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/py_rref.h:16:12: note:   no known conversion for argument 1 from 'std::shared_ptr<torch::distributed::rpc::UserRRef>' to 'c10::intrusive_ptr<torch::distributed::rpc::RRef>' 
Jan 28 01:09:28 /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/py_rref.h:15:12: note: candidate: torch::distributed::rpc::PyRRef::PyRRef(const pybind11::object&) 
Jan 28 01:09:28    explicit PyRRef(const py::object& value); 
Jan 28 01:09:28             ^ 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

zhaojuanmao committed:
Update on "[jit] Add RRef to IValue and Jit"


Differential Revision: [D19502233](https://our.internmc.facebook.com/intern/diff/D19502233/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[jit] Add RRef to IValue and Jit"


Differential Revision: [D19502233](https://our.internmc.facebook.com/intern/diff/D19502233/)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -16,6 +16,19 @@ void confirmPendingUser(
   ctx.delPendingUser(rr->forkId());
 }
+
+void finishCreatingOwnerRRef(
```
Add comment?

```
// Callback for RemoteCall to self.
void finishCreatingOwnerRRef(
```

mrshenli replied:
will need to add `TORCH_API` to expose it to invocations out of libtorch

zhaojuanmao commented:
@xush6528 thanks for reviewing the diff, but this diff is not ready to review yet, still need to clean up, thanks!

zhaojuanmao committed:
Update on "[jit] Add RRef to IValue and Jit"


Differential Revision: [D19502233](https://our.internmc.facebook.com/intern/diff/D19502233/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/core/rref_interface.h:
```diff
@@ -1,14 +1,16 @@
 // This abstract class contains only user-facing APIs, and will be shared
 // between jit and distributed to implement TorchScript support.
-class RRefInterface {
+class C10_EXPORT RRefInterface : public c10::intrusive_ptr_target {
```
what's the implication of inheriting from intrusive_ptr_target? Does this mean `RRefInterface` will always be a shared_ptr like object?

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -385,9 +385,19 @@ If the future completes with an error, an exception is thrown.
+        if (ctx.getWorkerId() != dstWorkerInfo.id_) {
+          auto userRRefPtr = remoteTorchscript(
+              dstWorkerName, qualifiedName, functionSchema, stack);
+          return PyRRef(userRRefPtr);
```
Does this mean, we can actually remove `PyRRef` type?

zhaojuanmao commented:
this diff is duplicated of #32992, please do not review this diff, I'm exporting it for testing purpose in the stacked diffs, will change title as well

zhaojuanmao committed:
Update on "[jit] Add RRef to IValue and Jit"


Differential Revision: [D19502233](https://our.internmc.facebook.com/intern/diff/D19502233/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[jit] Add RRef to IValue and Jit"


Differential Revision: [D19502233](https://our.internmc.facebook.com/intern/diff/D19502233/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[jit] Add RRef to IValue and Jit"


Differential Revision: [D19502233](https://our.internmc.facebook.com/intern/diff/D19502233/)

[ghstack-poisoned]

zzzwen commented:
Can we add Python3 type support 
`List[rpc.RRef[torch.Tensor]]` 
Will produce error 
`TypeError: 'pybind11_type' object is not subscriptable`

zhaojuanmao commented:
#32992 was landed
